### DISCUSSION TITLE
Can I get webgpu capabilities through nodejs?

### DISCUSSION THREAD

**Discussion by wewindy** [nodejs/help#4585] [Q&A] [answered]
As the title.

``` js
// main.js
const entryFunc = async () => {
  const adapter = await navigator.gpu.requestAdapter()
  const device = await adapter.requestDevice()

  console.log(device.limits)
  // ...
}
```

``` sh
node ./main.js
```

**Comment by targos** [comment, marked as answer]
WebGPU is not implemented in Node.js.

Maybe we could integrate it using the C++ library that is being made for Chromium: https://dawn.googlesource.com/dawn

**Reply by wewindy** [reply]
Well so, is there any plan to implement webgpu in nodejs?

**Reply by targos** [reply]
Not for now.

**Reply by wewindy** [reply]
Thank you!

**Reply by richardlau** [reply]
You could try Electron https://github.com/electron/electron/issues/26944

**Comment by robertleeplummerjr** [comment]
https://github.com/maierfelix/webgpu could be of use.  Perhaps we could ask to turn into organization and continue support?:

**Reply by targos** [reply]
The Dawn project seems to be actively working on Node.js bindings: https://dawn.googlesource.com/dawn/+/refs/heads/main/src/dawn/node/

**Reply by robertleeplummerjr** [reply]
![](https://media2.giphy.com/media/udmx3pgdiD7tm/giphy.gif?cid=ecf05e47zrz6mievz2uy5nknzicj2oztbd42wgh8o65nm9bx&rid=giphy.gif&ct=g)

**Reply by axinging** [reply]
Based on dawn's nodejs, I sucessfully run some examples on Windows/Linux:
Example: webgpu on node
https://github.com/axinging/webgpu-node-demo/tree/main/external-node

Example: tfjs-webgpu on node
https://github.com/axinging/webgpu-node-demo/tree/main/tfjs-on-external-node

**Reply by ovflowd** [reply]
@targos could we add these bindings to the feature requests board of Node.js? Feels like something that could be added im the future of Node.js

**Reply by bnoordhuis** [reply]
There's https://github.com/nodejs/node/issues/42896 already but it was closed due to inactivity.

WebGPU is still niche enough at this point that we can punt on it for another year and see what the uptake is.

**Reply by loretoparisi** [reply]
Why wait considering WebGPU is now in beta in Chrome and the raise of recent Chat LLM that can run directly in JavaScript thanks to the WebLLM project?
Let's also consider the recent interest in a TypeScript/JavaScript pure Tensor library.

**Reply by earonesty** [reply]
no longer beta.  it's live in chrome.

**Comment by benjamingr** [comment]
There are official Node.js bindinging by the Chromium team IIRC

**Reply by benjamingr** [reply]
https://dawn.googlesource.com/dawn/+/refs/heads/main/src/dawn/node/ found it, looks like it's under active development

**Reply by benjamingr** [reply]
Also cc @Kangz

**Reply by Kangz** [reply]
The Dawn team uses these node bindings for its own testing so we maintain them. They aren't exactly production ready yet, but should be pretty decent. The goal is to eventually have automated builds for them that are published on NPM but we didn't do that yet.

**Reply by kungfooman** [reply]
Any update for 2024? This is highly interesting for neural network libraries (currently testing diffusion.js in browser). Would be lovely to have the same API and have it working in nodejs aswell.

**Reply by greggman** [reply]
There's a build of dawn.node here

https://npmjs.org/webgpu
